### PR TITLE
Flush buffer after parsing escapes

### LINKED ISSUE #113: valid json reported as invalid
The file below contains a valid json string. However, the boost json example program `validate` reports it is invalid, and the example program `pretty` reports `caught exception: extra data`. 

[valid_json_boost_json_issue.txt](https://github.com/CPPAlliance/json/files/4803279/valid_json_boost_json_issue.txt)

Many small changes to the string makes it parse correctly, such as deleting an innocuous looking character from the string.

I've reproduced this on the tip of master (f47fbeb1a54b832221a35c17912febb7a2c3a317) with both clang and gcc.

Please let me know any additional information I can provide that would be helpful.

### PR DESCRIPTION
**Pull request by sdkrystian**
Fixes #113

### CODE REVIEW

File: include/boost/json/basic_parser.hpp [hunk 1]
```diff
@@ -734,30 +734,32 @@ parse_string(
         start = cs.data();
         state st;
         st_.pop(st);
-        switch(st)
+        if (st != state::str1)
```
Comment (vinniefalco) [inline include/boost/json/basic_parser.hpp:737]: This is now irregular (different from all the other switch-restore stanzas in the file).

File: include/boost/json/basic_parser.hpp [hunk 2]
```diff
@@ -807,6 +809,13 @@ parse_string(
                             return result::fail;
                     }
                 }
+                // KRYSTIAN TODO: this check is performed again
+                // at the label. Clip & increment the stream, then jump to
+                // do_str3
+
+                // JSON escapes can never make the
+                // transcoded utf8 string larger.
+                cs.clip(temp.max_size());
```
Comment (vinniefalco) [inline include/boost/json/basic_parser.hpp:818]: Why `max_size` instead of `capacity`?

File: include/boost/json/basic_parser.hpp [hunk 3]
```diff
@@ -1094,30 +1100,33 @@ parse_string(
             ++cs;
             continue;
         }
-        if(BOOST_JSON_LIKELY(more_))
+
+        // flush if the stream is empty
```
Comment (vinniefalco) [inline include/boost/json/basic_parser.hpp:1104]: LOL the comment is not only unhelpful, but wrong. It should say "buffer" not "stream", and it should say "not empty."

File: include/boost/json/basic_parser.hpp [hunk 4]
```diff
@@ -1094,30 +1100,33 @@ parse_string(
...
-            suspend(state::str3);
+            else
+            {
+                if(BOOST_JSON_UNLIKELY(
+                    ! h.on_string_part(temp, ec_)))
+                    return result::fail;
+            }
+            temp.clear();
         }
+        // if the stream was empty because it was clipped,
```
Comment (vinniefalco) [inline include/boost/json/basic_parser.hpp:1122]: is empty

File: test/basic_parser.cpp [hunk 5]
```diff
@@ -778,6 +778,41 @@ class basic_parser_test
         BOOST_TEST(! validate(js));
     }
 
+    void
+    testIssue113()
+    {
+        string_view s = 
+            "\"\\r\\n section id='description'>\\r\\nAll        mbers form the uncountable set "
```
Comment (vinniefalco) [inline test/basic_parser.cpp:785]: If this is from the issue can we please use a raw string so we don't have to have all these escapes?

File: include/boost/json/basic_parser.hpp (hunk 3 above)
Comment (sdkrystian) [reply include/boost/json/basic_parser.hpp:1104]: Hah, this proves exactly why this comment is needed. If you look at the `if` statement above, if tests `*cs`, meaning that this statement is only executed when the stream is empty :)

File: test/basic_parser.cpp (hunk 5 above)
Comment (sdkrystian) [reply test/basic_parser.cpp:785]: I used escapes to ensure that the strings are identical (i.e. no extra newlines)

Comment (sdkrystian) [reply test/basic_parser.cpp:785]: I used escapes to ensure that the strings are identical (i.e. no extra newlines)

File: include/boost/json/basic_parser.hpp (hunk 1 above)
Comment (sdkrystian) [reply include/boost/json/basic_parser.hpp:737]: The alternative is the keep `state::str1` and manually call trim at each case label

**Pull request closed without merging**